Add security patterns page (#2044)
sourcegraph/about: handbook/engineering/security_patterns.md

# Security patterns

## Authorization

> Authorization is the gatekeeper of securing private resources by preventing unauthorized access to those private resources. Examples of private resources are: 

Reviewer: ```suggestion
Authorization is the gatekeeper for securing private resources by preventing unauthorized access. Examples of private resources are: 
```

> - Content of a private repository

Reviewer: ```suggestion
- Contents of a private repository
```

- Endpoints that can only be called by certain users (e.g. site admins)

> - Settings of a user or an organization

Reviewer: ```suggestion
- User settings or organization settings
```

> Current forms of authorization in Sourcegraph including site-admin roles, organization memberships, campaigns permissions, repository permissions, and same user validation.

Reviewer: ```suggestion
Current forms of authorization in Sourcegraph include:

 - Site admin roles
 - Organization memberships
 - Campaigns permissions
 - Repository permissions
 - Same-user validation
```

> As a standard practice, users that do not have access to certain private resources should not be aware of the existencce of those private resources. In the case where only part of resources are restricted by authorization, it is reasonable to prompt the user that some resources are not avaialble because of insufficent permissions. However, in any case, we must not provide anything that can be used to indicate what those restricted resources might be.

Reviewer: ```suggestion
As a standard practice, users who do not have access to a given private resource should not be aware of the existence of that private resource. When only part of a resource is restricted by authorization limitations, it is reasonable to prompt the user that some resources are not available due to insufficient permissions. However, we must not provide anything that indicates what those restricted resources might be.
```

### Enforce authorization

> In Sourcegraph, there are two types of places to enforce authorization, both of them are equally important:

Reviewer: ```suggestion
In Sourcegraph, there are two places to enforce authorization, both equally important:
```

- At the GraphQL layer:
    - Some endpoints are restricted to certain users (e.g. site admins or the same user).

>     - Any type of backend failure may indicate information about private resources that are not authorized to the client. Therefore, we should halt the process as soon as we know the request is unauthorized, and behave as if the resource does not exist at all.

Reviewer: ```suggestion
    - Be aware that any backend failure has potential to indicate unauthorized information about private resources. Therefore, halt the process as soon as we identify an unauthorized request, and behave as if the resource does not exist at all.
```

- At the database layer:

>     - It is our source of truth for authorization, especially for repository and campaigns permissions. Enforcing authorization at this layer is absolutely necessary.

Reviewer: ```suggestion
    - The database our source of truth for authorization, especially for repository and campaigns permissions. Enforcing authorization at this layer is absolutely necessary.
```

You'll have to confirm this change is technically correct – is the source of truth the "database" or the "database layer?"
